hw/dv/dpi/dmidpi/dmidpi.c

```diff
@@ -0,0 +1,410 @@
+      } else {
+        ctx->jtag.jtag_state = PauseDr;
+      }
+      return false;
+    case PauseDr:
+      if (tms) {
+        ctx->jtag.jtag_state = Exit2Dr;
+      }
+      return false;
+    case Exit2Dr:
+      if (tms) {
+        ctx->jtag.jtag_state = UpdateDr;
+      } else {
+        ctx->jtag.jtag_state = ShiftDr;
+      }
+      return false;
+    case UpdateDr:
+      ctx->jtag.dr_captured = ctx->jtag.dr_shift_reg;
+      if (tms) {
+        ctx->jtag.jtag_state = SelectDrScan;
+      } else {
+        ctx->jtag.jtag_state = RunTestIdle;
+      }
+      // If a DMI read or write completes, write it out
+      if ((ctx->jtag.ir_captured == DMIAccess) &&
+          ((ctx->jtag.dr_captured & 0x3) != 0)) {
+        issue_dmi_req(ctx);
+        return true;
+      }
+      return false;
+    case SelectIrScan:
+      if (tms) {
+        ctx->jtag.jtag_state = TestLogicReset;
+      } else {
+        ctx->jtag.jtag_state = CaptureIr;
+      }
+      return false;
+    case CaptureIr:
+      ctx->jtag.ir_shift_reg = 0x5;
+      if (tms) {
+        ctx->jtag.jtag_state = Exit1Ir;
+      } else {
+        ctx->jtag.jtag_state = ShiftIr;
+      }
+      return false;
+    case ShiftIr:
+      ctx->jtag.ir_shift_reg |= ((uint32_t)tdi << 5);
+      ctx->jtag.ir_shift_reg >>= 1;
+      if (tms) {
+        ctx->jtag.jtag_state = Exit1Ir;
+      }
+      return false;
+    case Exit1Ir:
+      if (tms) {
+        ctx->jtag.jtag_state = UpdateIr;
+      } else {
+        ctx->jtag.jtag_state = PauseIr;
+      }
+      return false;
+    case PauseIr:
+      if (tms) {
+        ctx->jtag.jtag_state = Exit2Ir;
+      }
+      return false;
+    case Exit2Ir:
+      if (tms) {
+        ctx->jtag.jtag_state = UpdateIr;
+      } else {
+        ctx->jtag.jtag_state = ShiftIr;
+      }
+      return false;
+    case UpdateIr:
+      ctx->jtag.ir_captured = ctx->jtag.ir_shift_reg;
+      if (tms) {
+        ctx->jtag.jtag_state = SelectDrScan;
+      } else {
+        ctx->jtag.jtag_state = RunTestIdle;
+      }
+      return false;
+  }
+  return false;
+}
+
+/**
+ * Process a new command byte
+ *
+ * @param ctx a dmi context object
+ * @param cmd a packaged OpenOCD cmd
+ * @return true when a command completes, false otherwise
+ */
+static bool process_cmd_byte(struct dmidpi_ctx *ctx, char cmd) {
+  bool act_send_resp = false;
+  bool act_quit = false;
+
+  /*
+   * Documentation pointer:
+   * The remote_bitbang protocol implemented below is documented in the OpenOCD
+   * source tree at doc/manual/jtag/drivers/remote_bitbang.txt, or online at
+   * https://repo.or.cz/openocd.git/blob/HEAD:/doc/manual/jtag/drivers/remote_bitbang.txt
+   */
+
+  // parse received command byte
+  if (cmd >= '0' && cmd <= '7') {
+    // JTAG write
+    char cmd_bit = cmd - '0';
+    char tdi = (cmd_bit >> 0) & 0x1;
+    char tms = (cmd_bit >> 1) & 0x1;
+    char tck = (cmd_bit >> 2) & 0x1;
+    return (process_jtag_cmd(ctx, tdi, tms, tck));
```

Comment:
Please either use the `act_*` signals to indicate an action, and execute this action at the bottom of this function. This will require two more actions for the read and reset actions. Or insert the actions as function calls in the if/else blocks (as for the JTAG write); but then remove the other two `act_quit` and `act_send_resp` and inline them as well (that's probably nicer).

Reply:
In addition to Philipp's suggsetions, I would order the commands as in the referenced doc.